pkg/reconciler/gc/v2/gc.go

```diff
@@ -61,51 +64,66 @@ func Collect(
 		return a.After(b)
 	})
 
-	for _, rev := range revs[gcSkipOffset:] {
-		if strings.EqualFold(rev.ObjectMeta.Annotations[serving.RevisionPreservedAnnotationKey], "true") {
+	stale, nonactive := 0, 0
+	for _, rev := range revs {
+		if isRevisionActive(rev, config) {
+			continue
+		}
+		nonactive++
+		if isRevisionStale(cfg, rev, logger) {
+			stale++
+		}
+
+		if stale > min {
+			logger.Info("Deleting stale revision: ", rev.ObjectMeta.Name)
```

Comment:
I wonder if we should tweak the name of `MinNonActiveRevisions` and `MaxNonActiveRevisions` config options for more clarity
 
https://github.com/knative/serving/blob/d86ab4a68ec6421396711f32321618f62456836e/pkg/gc/config.go#L58-L63


Since depending on how the revisions are processes you'd get different gc results

For example (with min 20 & max 1000):
A list composed of 20 stale revisions and then 100 non-active would cause all non-active revisions to be deleted
compared to
A list composed of 100 non-active revisions and then 20 stale revisions would delete nothing

Unsure what we want

Reply:
The intention is the latter.

I think I've documented that better here: https://github.com/knative/serving/blob/da8105f8c1bec37f9302309e045fb01a805c835f/config/core/configmaps/gc.yaml#L59
and need to carry that across to the comment. Note that all stale revisions are also non-active. In your example I think we'd expect 20 to be deleted and the remaining to stick around until they also go stale. However if you disable the time bounds for staleness, we keep growing the revision list to the max. That allows you to either have a hard-count of how many we keep around or a less deterministic time-based approach that can still have a ceiling. Or disable everything!

Reply:
I think I misunderstood what you were saying. The latter scenario would be unexpected because the revisions are first sorted by their last active time. We are expecting stale revisions to show up first in the list, but let me try to think of an example where that assumption fails.

Reply:
I've updated the code to consider the latter scenario regardless of the sort assumption. I think it's a little more readable too.